includes/class-newspack-newsletters-subscription.php

```diff
@@ -402,8 +414,187 @@ public static function add_contact( $contact, $lists = false ) {
+			return $intents;
+		}
+		return $intents[ wp_hash( $email ) ] ?? false;
+	}
+
+	/**
+	 * Remove subscription intent.
+	 *
+	 * @param string $email Email address.
+	 *
+	 * @return void
+	 */
+	private static function remove_subscription_intent( $email ) {
+		$intents = get_option( 'newspack_newsletters_subscription_intents', [] );
+		if ( ! is_array( $intents ) ) {
+			$intents = [];
+		}
+		unset( $intents[ wp_hash( $email ) ] );
+		update_option( 'newspack_newsletters_subscription_intents', $intents );
+	}
+
+	/**
+	 * Process subscription intent.
+	 *
+	 * @param string $email Optional email address to process. It'll process all
+	 *                      intents if email is not provided.
+	 *
+	 * @return void
+	 */
+	public static function process_subscription_intents( $email = null ) {
+		if ( $email ) {
+			$intents = [ self::get_subscription_intent( $email ) ];
+		} else {
+			$intents = get_option( 'newspack_newsletters_subscription_intents', [] );
+		}
+		foreach ( $intents as $intent ) {
+			if ( empty( $intent ) ) {
+				continue;
+			}
+			if ( count( $intent['errors'] ) > 3 ) {
+				Newspack_Newsletters_Logger::log( 'Too many errors for contact ' . $intent['contact']['email'] . '. Removing intent.' );
+				self::remove_subscription_intent( $intent['contact']['email'] );
+				continue;
+			}
+			$contact     = $intent['contact'];
+			$lists       = $intent['lists'];
+			$is_updating = $intent['is_updating'];
+			$result      = self::add_contact_to_provider( $contact, $lists, $is_updating );
+
+			$user = get_user_by( 'email', $email );
+			if ( \is_wp_error( $result ) ) {
+				$email = $contact['email'];
+				if ( $user ) {
+					update_user_meta( $user->ID, 'newspack_newsletters_subscription_error', $result->get_error_message() );
+				}
+				$intent['errors'][] = $result->get_error_message();
+				Newspack_Newsletters_Logger::log( 'Error adding contact: ' . $result->get_error_message() );
+			} else {
+				self::remove_subscription_intent( $email );
+				if ( $user ) {
+					delete_user_meta( $user->ID, 'newspack_newsletters_subscription_error' );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Get a user subscription intent error.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return string|false Error message or false if not found.
+	 */
+	private static function get_user_subscription_intent_error( $user_id ) {
+		return get_user_meta( $user_id, 'newspack_newsletters_subscription_error', true );
+	}
+
+	/**
+	 * Schedule subscription intents processing.
+	 */
+	public static function schedule_subscription_intents_processing() {
+		if (
+			! empty( get_option( 'newspack_newsletters_subscription_intents', [] ) ) &&
+			! \wp_next_scheduled( 'newspack_newsletters_process_subscription_intents' )
+		) {
+			\wp_schedule_single_event( time() + 60, 'newspack_newsletters_process_subscription_intents' );
+		}
+		add_action( 'newspack_newsletters_process_subscription_intents', [ __CLASS__, 'process_subscription_intents' ] );
```

Comment:
When you do this, aren't you open the possibility that this callback will run in any random request. So this could delay a random user from seeing the site if they are unlucky enough. I know this should be an edge case because we will only fall here if there were errors in the async request. Thoughts?

(just curious, any difference between using single_event and re-scheduling it to 60 seconds in the future and using recurring events?)

Reply:
Just thinking that this has a potential to hurt a site in case there's an issue in the ESP.

If the ESP API has an issue for a moment, error might start to pile up.

My personal preference would be not to try again, and just send the user an email telling them about the error and asking them to try again (maybe with a handy link that would do it right away).

But it's not a blocker for me, I can see how you built some guard rails here... I'm gonna test everything

Reply:
> When you do this, aren't you open the possibility that this callback will run in any random request. So this could delay a random user from seeing the site if they are unlucky enough.

Depends on how the system leverages WP cron. With the default WP way, yes, but with appropriate caching in place, the edge case of an errored async intent should most likely be processed in an admin execution but could also hit a logged-in reader... Not sure there's any way around this, that's scheduling tasks in WP.

> (just curious, any difference between using single_event and re-scheduling it to 60 seconds in the future and using recurring events?)

There's no need to schedule the task if there are no intents in the db.

> My personal preference would be not to try again, and just send the user an email telling them about the error and asking them to try again (maybe with a handy link that would do it right away).

That's a good point, but sending emails to readers also comes with risks. RAS welcome emails were disabled because readers would often resolve subscription campaign prompts with junk email, which would impact the email server score negatively. I wouldn't rule this option out, though. I see the current implementation as a simple lower-risk approach that we can iterate over.

Reply:
> There's no need to schedule the task if there are no intents in the db.

Gotcha! Nice!